src/CSStrikes.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-FileCopyrightText: 2024 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.24;
+
+import { MerkleProof } from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import { AccessControlEnumerableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/extensions/AccessControlEnumerableUpgradeable.sol";
+import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import { ICSStrikes } from "./interfaces/ICSStrikes.sol";
+
+/// @author vgorkavenko
+contract CSStrikes is
+    ICSStrikes,
+    Initializable,
+    AccessControlEnumerableUpgradeable
+{
+    bytes32 public constant STRIKES_MANAGER_ROLE =
+        keccak256("STRIKES_MANAGER_ROLE");
+
+    address public immutable ORACLE;
+
+    /// @notice The latest Merkle Tree root
+    bytes32 public treeRoot;
+
+    /// @notice CID of the last published Merkle tree
+    string public treeCid;
+
+    constructor(address oracle) {
+        if (oracle == address(0)) revert ZeroOracleAddress();
+
+        ORACLE = oracle;
+
+        _disableInitializers();
+    }
+
+    function initialize(address admin) external initializer {
+        __AccessControlEnumerable_init();
+        if (admin == address(0)) revert ZeroAdminAddress();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+    }
+
+    /// @inheritdoc ICSStrikes
+    function processOracleReport(
+        bytes32 _treeRoot,
+        string calldata _treeCid
+    ) external {
+        if (msg.sender != ORACLE) revert NotOracle();
+
+        bool isSameCid = keccak256(bytes(_treeCid)) ==
+            keccak256(bytes(treeCid));
+        bool isSameRoot = _treeRoot == treeRoot;
+
+        if (bytes(_treeCid).length != 0 && _treeRoot != bytes32(0)) {
+            if (isSameCid) revert InvalidTreeCID();
+            if (isSameRoot) revert InvalidTreeRoot();
+        }
+
+        if (!isSameCid && !isSameRoot) {
```

Comment:
that's a pretty strange condition, don't get it
it allows to ignore setting if any of these are the same, but it seems to be an illegal values, isn't it? in which cases we will receive such values?

Reply:
We don't want to update values from 0 to 0. But in the same time the new one can be empty  when the old one not

Reply:
> If you have the same root, you should have the same CID, right?

Yes

Reply:
so why do we have this condition here then

Reply:
I would prefer to keep it as is to be more solid and don't allow setting invalid data